### DISCUSSION TITLE
Question: How can create text index for json object?

### DISCUSSION THREAD

**Discussion by niklasgrewe** [typegoose/typegoose#413] [Q&A] [answered]
Hi, i want to create a text index for a json object. The json object looks like this:

```json
"content": {
    "time": 1550476186479,
    "blocks": [
        {
            "type": "header",
            "data": {
                "text": "Editor.js",
                "level": 2
            }
        },
        {
            "type": "paragraph",
            "data": {
                "text": "Hey. Meet the new Editor. On this page you can see it in action — try to edit this text. Source code of the page contains the example of connection and configuration."
            }
        }
    ],
    "version": "2.8.1"
},
```
my `.model.ts` file looks like this:

```js
@Index({title: 'text', description: 'text'})
export class Page {
    @prop({required: true})
    title: string;

    @prop({required: true})
    description: string;

    @prop({required: true})
    content: Content;

    @prop()
    createdAt?: Date;

    @prop()
    updatedAt?: Date;

    @prop()
    category: string;
}
```
I already created text indexes for the fields `title` and `description`. Everthing works as i aspected, but how can i set a text index for each `content.blocks.data.text` field so i can also search my content created with Editor.js? 

## Versions

- typegoose: 7.4.2
- mongoose: 5.10.15

**Comment by hasezoey** [comment]
if `Content` is an class (and the underlying fields too), then you can just use `@Index` on those, or maybe use dottet notation (i dont know if it will work), like `'content.blocks.data.text': 'text'`

**Comment by niklasgrewe** [comment]
@hasezoey `Content` is a class and looks like this
```js
export class Content {
    time: number;
    blocks: [
        {
            type: string,
            data: {
                text: string,
                level: string,
            },
        }
    ];
    version: string;
}
```
but when i set
```js
@Index({title: 'text', description: 'text', 'content.blocks.data.text': 'text'})
```
unfortunately it doesn't work. Any ideas?
Or could it not work because `blocks` is an array?

**Comment by hasezoey** [comment]
the problem seems to be that you dont add `@prop` to your sub-classes properties (which is required if you dont just want the types at type/compile-time, but not at runtime)

**Comment by niklasgrewe** [comment]
thanks i changed my `Content` class to this
```js
import { modelOptions, prop, Severity } from "@typegoose/typegoose";

@modelOptions({
    options: {
        allowMixed: Severity.ALLOW
    }
})
export class Content {
    @prop()
    time: number;

    @prop()
    blocks: [
        {
            type: string,
            data: {
                text: string,
                level: string,
            },
        }
    ];

    @prop()
    version: string;
}
```
but when i set `@Index()` to
```js
@Index({title: 'text', description: 'text', 'content.blocks.data.text': 'text'})
```
or to
```js
@Index({title: 'text', description: 'text', content: 'text'})
```
it doesn't work. I get zero results. 
What do i need to change too to get this working?

**Comment by hasezoey** [comment]
did you already try making the array contents an class (and data as well) into their own class?

did you always make an new collection or ran `Model.syncIndexes()`?

**Comment by niklasgrewe** [comment]
neither nor... I'm honestly still a beginner when it comes to backend development. Could you please show me what you mean and what I should change?

**Comment by hasezoey** [comment, marked as answer]
if the collection already exists, then mongoose dosnt create new indexes in it without removing the collection and recreating it, or running `Model.syncIndexes()` (deletes indexes that are not on the model and adds the ones that are not in the collection)

**Comment by niklasgrewe** [comment]
thank you for the explanation I did not know that. With `runSyncIndexes` it works now. 👍 Should the function in production mode be removed?

**Comment by hasezoey** [comment]
you can let it in, but if you should ever change the indexes (on an existing model) it will be updated right after the next restart, this might sometimes be unwanted behaviour - if forgotten to remove it and accidentally change an index

**Comment by niklasgrewe** [comment]
ok thank you so the case can be closed now
